netty/src/main/java/io/grpc/netty/NettyClientStream.java

```diff
@@ -103,47 +101,55 @@ protected Sink abstractClientStreamSink() {
 
   @Override
   public void setAuthority(String authority) {
-    checkState(headers != null, "must be call before start");
     this.authority = AsciiString.of(checkNotNull(authority, "authority"));
   }
 
   @Override
-  public void start(ClientStreamListener listener) {
-    super.start(listener);
+  public Attributes getAttributes() {
+    return state.handler.getAttributes();
+  }
 
-    // Convert the headers into Netty HTTP/2 headers.
-    AsciiString defaultPath = (AsciiString) methodDescriptorAccessor.geRawMethodName(method);
-    if (defaultPath == null) {
-      defaultPath = new AsciiString("/" + method.getFullMethodName());
-      methodDescriptorAccessor.setRawMethodName(method, defaultPath);
-    }
-    headers.discardAll(GrpcUtil.USER_AGENT_KEY);
-    Http2Headers http2Headers
-        = Utils.convertClientHeaders(headers, scheme, defaultPath, authority, userAgent);
-    headers = null;
+  private static boolean useGet(MethodDescriptor<?, ?> method) {
+    return method.isIdempotent() && method.isSafe()
+        && (method.getType() == MethodDescriptor.MethodType.UNARY);
+  }
 
-    ChannelFutureListener failureListener = new ChannelFutureListener() {
-      @Override
-      public void operationComplete(ChannelFuture future) throws Exception {
-        if (!future.isSuccess()) {
-          // Stream creation failed. Close the stream if not already closed.
-          Status s = transportState().statusFromFailedFuture(future);
-          transportState().transportReportStatus(s, true, new Metadata());
-        }
+  private class Sink implements AbstractClientStream2.Sink {
+    @Override
+    public void writeHeaders(Metadata headers, byte[] requestPayload) {
+      // Convert the headers into Netty HTTP/2 headers.
+      AsciiString defaultPath = (AsciiString) methodDescriptorAccessor.geRawMethodName(method);
+      if (defaultPath == null) {
+        defaultPath = new AsciiString("/" + method.getFullMethodName());
+        methodDescriptorAccessor.setRawMethodName(method, defaultPath);
+      }
+      headers.discardAll(GrpcUtil.USER_AGENT_KEY);
+      boolean get = (requestPayload != null);
+      if (get) {
+        // Forge the query string
```

Comment:
Optional: consider moving the assignment of the method along with assignment of the path